src/DefinitionResolver.php

```diff
@@ -667,13 +666,27 @@ public function resolveExpressionNodeToType($expr)
                 } else {
                     $classFqn = substr((string)$t->getFqsen(), 1);
                 }
-                $fqn = $classFqn . '->' . $expr->memberName->getText($expr->getFileContents());
+                $add = '->' . $expr->memberName->getText($expr->getFileContents());
                 if ($expr->parent instanceof Node\Expression\CallExpression) {
-                    $fqn .= '()';
+                    $add .= '()';
                 }
+                $fqn = $classFqn . $add;
                 $def = $this->index->getDefinition($fqn);
                 if ($def !== null) {
                     return $def->type;
+                } else {
+                    $classDef = $this->index->getDefinition($classFqn);
```

Comment:
> Is there a way to mark the initial type properly as Types\This, and then resolve the type at the point where it is accessed?
I can not think of a way to do this, as sometimes the type is returned right away - without further processing. So if I am in the scope of ```CompletionProvider``` and get a ```Type\This``` node there is nothing I can do with it at that point. But at the same time - resolving a node to type may cause a lot of recursion - in which case ```Type\This``` may be useful until the value is returned to the consumer. To be honest, provided there is no performance hit, my approach would be to convert the type to something concrete as soon as possible (as opposed to dealing with a meta type like ```This```).